app/controllers/RestApi.scala

```diff
@@ -87,6 +101,118 @@ class RestApi @Inject() (
+     placeOfResidence: String,
+     birthday: Long,
+     sex: String,
+     password: String,
+     profileImageUrl: Option[String]
+  )
+  object CreateUserBody{
+    implicit val createUserBodyJsonFormat = Json.format[CreateUserBody]
+  }
+
+  def createUser() = ApiAction.async(validateJson[CreateUserBody]) { implicit request => {
+    val signUpData = request.request.body
+
+    val loginInfo = LoginInfo(CredentialsProvider.ID, signUpData.email)
+    userService.retrieve(loginInfo).flatMap{
+      case Some(_) =>
+        Future(BadRequest(Json.obj("error" -> Messages("error.userExists", signUpData.email))))
+      case None =>{
+        val profile = Profile(loginInfo, true, signUpData.email, signUpData.firstName, signUpData.lastName, signUpData.mobilePhone, signUpData.placeOfResidence, signUpData.birthday, signUpData.sex, List(new DefaultProfileImage))
+        avatarService.retrieveURL(signUpData.email).flatMap(avatarUrl  => {
+          userService.save(User(id = UUID.randomUUID(), profiles =
+            (signUpData.profileImageUrl, avatarUrl) match {
+              case (Some(url), Some(gravatarUrl))  => List(profile.copy(avatar = List(UrlProfileImage(url),GravatarProfileImage(gravatarUrl),new DefaultProfileImage)))
+              case (None, Some(gravatarUrl))=> List(profile.copy(avatar = List(GravatarProfileImage(gravatarUrl),new DefaultProfileImage)))
+              case (Some(url), None) => List(profile.copy(avatar = List(UrlProfileImage(url), new DefaultProfileImage)))
+              case _ => List(profile.copy(avatar = List(new DefaultProfileImage)))
+            })).flatMap((user) => {
+              authInfoRepository.add(loginInfo, passwordHasher.hash(signUpData.password)).map(_ => Ok(Json.toJson(user)))
+          })
+        })
+      }
+    }
+  }}
+
+  case class UpdateUserBody(
+                             email: String,
+                             firstName : Option[String],
+                             lastName: Option[String],
+                             mobilePhone: Option[String],
+                             placeOfResidence: Option[String],
+                             birthday: Option[Long],
+                             sex: Option[String],
+                             password: Option[String]
+                           )
+  object UpdateUserBody {
+    implicit val updateUserBodyJsonFormat = Json.format[UpdateUserBody]
+  }
+
+  def updateUser() = ApiAction.async(validateJson[UpdateUserBody]){ implicit request =>{
+    val userData = request.request.body
+    val loginInfo : LoginInfo = LoginInfo(CredentialsProvider.ID, userData.email)
+    val supporter : Supporter = Supporter(userData.firstName, userData.lastName, userData.mobilePhone, userData.placeOfResidence, userData.birthday, userData.sex)
+    userDao.find(loginInfo).flatMap(userObj => {
+      userObj match {
+        case Some(user) => user.profileFor(loginInfo) match {
+          case Some(profile) => {profile.supporter.copy(
```

Comment:
The updated copy of supporter stays unused. It should be sved into a value "supporter".

Reply:
Nope, sorry. It's already done in line 160. Thus, the whole copy can be removed?